kyo-prelude/shared/src/main/scala/kyo/Pipe.scala

```diff
@@ -0,0 +1,721 @@
+                [C] =>
+                    (unit, cont) =>
+                        Poll.andMap[Chunk[A1]]:
+                            case Absent => Loop.continue(cont(Absent))
+                            case Present(chunk) =>
+                                Kyo.foreach(chunk)(f).map: chunk2 =>
+                                    Loop.continue(cont(Present(chunk2)))
+            )
+
+    /** Transform a pipe to consume a stream of a different element type using a pure mapping function that transforms streamed chunks.
+      *
+      * @param f
+      *   Pure function mapping chunks of a new stream element type to chunks of the original stream element type
+      * @return
+      *   A pipe that accepts streams of the new element type
+      */
+    def contramapChunk[A1](f: Chunk[A1] => Chunk[A])(
+        using
+        t1: Tag[Poll[Chunk[A]]],
+        t2: Tag[A1],
+        fr: Frame
+    ): Pipe[A1, B, S] =
+        Pipe:
+            ArrowEffect.handleLoop(t1, pollEmit)(
+                [C] =>
+                    (unit, cont) =>
+                        Poll.andMap[Chunk[A1]]: maybeChunk =>
+                            Loop.continue(cont(maybeChunk.map(f)))
+            )
+
+    /** Transform a pipe to consume a stream of a different element type using an effectful mapping function that transforms streamed
+      * chunks.
+      *
+      * @param f
+      *   Effectful function mapping chunks of a new stream element type to chunks of the original stream element type
+      * @return
+      *   A pipe that accepts streams of the new element type
+      */
+    def contramapChunk[A1, S1](f: Chunk[A1] => Chunk[A] < S1)(
+        using
+        t1: Tag[Poll[Chunk[A]]],
+        t2: Tag[A1],
+        disc: Discriminator,
+        fr: Frame
+    ): Pipe[A1, B, S & S1] =
+        Pipe:
+            ArrowEffect.handleLoop(t1, pollEmit)(
+                [C] =>
+                    (unit, cont) =>
+                        Poll.andMap[Chunk[A1]]:
+                            case Absent => Loop.continue(cont(Absent))
+                            case Present(chunk) =>
+                                f(chunk).map: chunk2 =>
+                                    Loop.continue(cont(Present(chunk2)))
+            )
+
+    /** Transform a pipe to produce a new output type using a pure function that transforms each streamed element of the original pipe's
+      * output stream.
+      *
+      * @param f
+      *   Pure function mapping the streamed elements produced by the original pipe
+      * @return
+      *   A new pipe with output stream transformed by the mapping function
+      */
+    def map[B1](f: B => B1)(
+        using
+        t1: Tag[Emit[Chunk[B]]],
+        t2: Tag[B1],
+        fr: Frame
+    ): Pipe[A, B1, S] =
+        Pipe:
+            ArrowEffect.handleLoop(t1, pollEmit)(
+                [C] =>
+                    (chunk, cont) =>
+                        Emit.valueWith(chunk.map(f))(Loop.continue(cont(())))
```

Comment:
Is there a way we could avoid the code duplication with `Stream`? Looking at [Stream.map](https://github.com/getkyo/kyo/blob/0e91844fa55cb8b32289d3283965aa1f673fa8e8/kyo-prelude/shared/src/main/scala/kyo/Stream.scala#L67), it seems this code is missing avoiding emits in case the chunk is empty. We might end up with multiple mismatches between the implementations. Maybe internal APIs that are used by both `Stream` and `Pipe` to avoid the `Pipe` allocation? I wonder if we're missing an effect in `kyo-prelude`.

Reply:
Thank you for the elaboration! I'd be ok with some tech debt due to code duplication. What do you think is the best path? With the 1.0-RC cycle coming up, I'd say it's defining the APIs we want to have is more relevant than accumulating some tech debt in in internal code. Performance is similar, it's a major focus of the project but we can  typically optimize things later